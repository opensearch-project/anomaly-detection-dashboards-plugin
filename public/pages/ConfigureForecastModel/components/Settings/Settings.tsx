/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React, { Fragment, useEffect, useRef, useState } from 'react';
import { Field, FieldProps, FormikProps } from 'formik';
import {
  EuiSpacer,
  EuiFlexGroup,
  EuiFlexItem,
  EuiBadge,
  EuiCompressedFieldNumber,
  EuiRange,
  EuiText,
  EuiPanel,
} from '@elastic/eui';
import {
  isInvalid,
  getError,
  validatePositiveInteger,
  validateNonNegativeInteger,
  validateHistory,
} from '../../../../utils/utils';
import { FORECASTER_DOCS_LINK, FIELD_MAX_WIDTH, INPUT_SLIDER_WIDTH } from '../../../../utils/constants';
import { FormattedFormRow } from '../../../../components/FormattedFormRow/FormattedFormRow';
import '../../index.scss';
import { toNumberOrEmpty } from '../../utils/helpers';

interface SettingsProps {
  isEditable?: boolean;
}

export const Settings = ({ isEditable = true }: SettingsProps) => {
  return (
    <Fragment>
      {/* FORECASTING INTERVAL */}
      <Field name="interval" validate={validatePositiveInteger}>
        {({ field, form }: FieldProps<number>) => {
          const value = field.value ?? '';
          return (
            <FormattedFormRow
              fullWidth
              title="Forecasting interval"
              hint={['How often the forecast runs to generate next value.']}
              hintLink={FORECASTER_DOCS_LINK}
              isInvalid={isInvalid(field.name, form)}
              error={getError(field.name, form)}
              // FIXME: We are going to lift the 1 hour interval restriction.
              helpText="The interval must be at least one minute."
            >
              <EuiFlexGroup
                gutterSize="none"
                alignItems="center"
                style={{ maxWidth: FIELD_MAX_WIDTH }}
              >
                <EuiFlexItem grow={false}>
                  <div style={{ width: INPUT_SLIDER_WIDTH }}>
                    <EuiCompressedFieldNumber
                      name={field.name}
                      id={field.name}
                      value={value}
                      style={{ width: '100%' }}
                      disabled={!isEditable}
                      onChange={(e) => {
                        form.setFieldValue(field.name, toNumberOrEmpty(e.target.value));
                      }}
                    />
                  </div>
                </EuiFlexItem>
                <EuiFlexItem grow={false}>
                  <EuiBadge color="default" className="unit-badge">
                    minutes
                  </EuiBadge>
                </EuiFlexItem>
              </EuiFlexGroup>
            </FormattedFormRow>
          );
        }}
      </Field>

      <EuiSpacer size="l" />

      {/* WINDOW DELAY */}
      <Field name="windowDelay" validate={validateNonNegativeInteger}>
        {({ field, form }: FieldProps<number>) => {
          const value = field.value ?? '';
          return (
            <FormattedFormRow
              fullWidth
              title="Window delay"
              hint="Specify a window of delay for a forecaster to fetch data, if you need to account for extra processing time."
              hintLink={FORECASTER_DOCS_LINK}
              isInvalid={isInvalid(field.name, form)}
              error={getError(field.name, form)}
            >
              <EuiFlexGroup
                gutterSize="none"
                alignItems="center"
                style={{ maxWidth: FIELD_MAX_WIDTH }}
              >
                {/* grow={false} in both the numeric field and unitprevents horizontal expansion,
                 keeping the numeric field at its natural width and allows the unit badge to sit directly next to it.
                 Otherwise, there is a gap between the numeric field and the unit badge. */}
                <EuiFlexItem grow={false}>
                  <div style={{ width: INPUT_SLIDER_WIDTH }}>
                    <EuiCompressedFieldNumber
                      name={field.name}
                      id={field.name}
                      value={value}
                      style={{ width: '100%' }}
                      disabled={!isEditable}
                      onChange={(e) => {
                        form.setFieldValue(field.name, toNumberOrEmpty(e.target.value));
                      }}
                    />
                  </div>
                </EuiFlexItem>
                <EuiFlexItem grow={false}>
                  <EuiBadge color="default" className="unit-badge">
                    minutes
                  </EuiBadge>
                </EuiFlexItem>
              </EuiFlexGroup>
            </FormattedFormRow>
          );
        }}
      </Field>

      <EuiSpacer size="l" />

      <Field name="horizon" validate={validatePositiveInteger}>
        {({ field, form }: FieldProps<number>) => {
          const [showHorizonDetails, setShowHorizonDetails] = useState(false);

          const interval = (form as FormikProps<any>).values.interval;
          const horizon = field.value ?? 0;
          const value = field.value ?? '';

          // Ref to the SLIDER area; if a click happens outside this ref, we close the slider
          const sliderPanelRef = useRef<HTMLDivElement>(null);

          // UseEffect that only attaches the "outside click" listener when the slider is open
          useEffect(() => {
            if (!showHorizonDetails) return;

            function handleOutsideClick(e: MouseEvent) {
              // If the click is not inside sliderPanelRef => close slider
              if (
                sliderPanelRef.current &&
                !sliderPanelRef.current.contains(e.target as Node)
              ) {
                setShowHorizonDetails(false);
              }
            }

            // Note we use capture = true to catch the event early. Could also omit if it works.
            document.addEventListener('mousedown', handleOutsideClick, true);
            return () => {
              document.removeEventListener('mousedown', handleOutsideClick, true);
            };
          }, [showHorizonDetails]);

          // 2) If user moves slider => update field
          function onSliderChange(newVal: number) {
            form.setFieldValue(field.name, newVal);
          }

          // For help text
          function getHorizonHelpText(h: number, i: number) {
            if (!i || !h) return 'A valid horizon is between 1 and 180.';
            const totalMinutes = h * i;
            const hours = Math.floor(totalMinutes / 60);
            const minutes = totalMinutes % 60;
            return (
              `${h} intervals = ${hours} hour${hours === 1 ? '' : 's'}${minutes ? ` ${minutes} minute${minutes === 1 ? '' : 's'}` : ''
              } if the forecasting interval is ${i} minutes. A valid horizon is between 1 and 180.`
            );
          }

          return (
            <FormattedFormRow
              fullWidth
              title="Horizon"
              hint={['How far the forecast extends into the future.']}
              hintLink={FORECASTER_DOCS_LINK}
              isInvalid={isInvalid(field.name, form)}
              error={getError(field.name, form)}
              helpText={
                !showHorizonDetails
                  ? 'Click the number field for more options.'
                  : getHorizonHelpText(horizon, interval)
              }
            >
              {/* 
          Constrain the entire row for consistency with other fields. 
          The numeric field + badge is outside the slider ref. 
        */}
              <div style={{ maxWidth: FIELD_MAX_WIDTH }}>
                <EuiFlexGroup gutterSize="none" alignItems="center">
                  <EuiFlexItem grow={false}>
                    {/* 3) onMouseDown to open the slider, with e.stopPropagation() */}
                    <div style={{ width: INPUT_SLIDER_WIDTH }}>
                      <EuiCompressedFieldNumber
                        min={1}
                        max={180}
                        id={field.name}
                        value={value}
                        style={{ width: '100%' }}
                        disabled={!isEditable}
                        // Use onMouseDown so we can stopPropagation
                        onMouseDown={(e) => {
                          // Prevent this click from bubbling to outside-click handler
                          e.stopPropagation();
                          setShowHorizonDetails(true);
                        }}
                        onChange={(e) => {
                          form.setFieldValue(field.name, toNumberOrEmpty(e.target.value));
                        }}
                      />
                    </div>
                  </EuiFlexItem>
                  <EuiFlexItem grow={false}>
                    <EuiBadge color="default" className="unit-badge">
                      intervals
                    </EuiBadge>
                  </EuiFlexItem>
                </EuiFlexGroup>

                {/* 
            Conditionally render the slider in a ref-wrapped div 
               so that clicks inside don't close the slider. 
          */}
                {showHorizonDetails && (
                  <>
                    <EuiSpacer size="s" />
                    <div ref={sliderPanelRef}>
                      <EuiPanel hasBorder hasShadow={false} paddingSize="m" style={{ width: INPUT_SLIDER_WIDTH }}>
                        <EuiRange
                          style={{ width: '100%' }}
                          min={1}
                          max={180}
                          step={1}
                          value={String(horizon)}
                          onChange={(e) => onSliderChange(Number(e.currentTarget.value))}
                          showLabels
                          showRange
                          levels={[
                            { min: 1, max: 40, color: 'success' },
                            { min: 40, max: 180, color: 'warning' },
                          ]}
                          aria-label="Horizon slider"
                        />
                        <EuiSpacer size="xs" />
                        <EuiText size="xs" color="subdued">
                          Recommended horizon is up to 40 intervals.
                        </EuiText>
                      </EuiPanel>
                    </div>
                  </>
                )}
              </div>
            </FormattedFormRow>
          );
        }}
      </Field>

      <EuiSpacer size="l" />

      {/* HISTORY */}
      <Field name="history" validate={validateHistory}>
        {({ field, form }: FieldProps<number>) => {
          const interval = (form as FormikProps<any>).values.interval;
          const value = field.value ?? '';

          function getHistoryHelpText(history: number, intv: number) {
            if (!intv || !history) return 'Minimum history: 40 intervals.';
            const totalMinutes = history * intv;
            const hours = Math.floor(totalMinutes / 60);
            const mins = totalMinutes % 60;
            return `${history} intervals = ${hours} hour${
              hours === 1 ? '' : 's'
            }${
              mins ? ` ${mins} minute${mins === 1 ? '' : 's'}` : ''
            } if the forecasting interval is ${intv} minutes. Minimum history: 40 intervals.`;
          }

          return (
            <FormattedFormRow
              fullWidth
              title="History"
              hint={['How far back the model looks for training data.']}
              hintLink={FORECASTER_DOCS_LINK}
              isInvalid={isInvalid(field.name, form)}
              error={getError(field.name, form)}
              helpText={getHistoryHelpText(field.value, interval)}
            >
              {/* gutterSize="none" ensures there's no default spacing between flex items */}
              <EuiFlexGroup
                gutterSize="none"
                alignItems="center"
                style={{ maxWidth: FIELD_MAX_WIDTH }}
              >
                
                <EuiFlexItem grow={false}>
                  <div style={{ width: INPUT_SLIDER_WIDTH }}>
                    <EuiCompressedFieldNumber
                      min={40}
                      max={10000}
                      name={field.name}
                      id={field.name}
                      value={value}
                      disabled={!isEditable}
                    style={{ width: '100%' }}
                    onChange={(e) => {
                      form.setFieldValue(field.name, toNumberOrEmpty(e.target.value));
                    }}
                  />
                  </div>
                </EuiFlexItem>
                <EuiFlexItem grow={false}>
                  <EuiBadge color="default" className="unit-badge">
                    intervals
                  </EuiBadge>
                </EuiFlexItem>
              </EuiFlexGroup>
            </FormattedFormRow>
          );
        }}
      </Field>
    </Fragment>
  );
};
