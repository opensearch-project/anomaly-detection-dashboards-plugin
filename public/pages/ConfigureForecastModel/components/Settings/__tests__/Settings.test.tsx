/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Settings } from '../Settings';

import { Formik } from 'formik';

describe('<Settings /> spec', () => {
  test('renders the component', () => {
    const { container } = render(
      <Formik initialValues={{ detectorName: '' }} onSubmit={jest.fn()}>
        {() => (
          <div>
            <Settings />
          </div>
        )}
      </Formik>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  test('shows error for empty interval when toggling focus/blur', async () => {
    const { queryByText, findByText, getByPlaceholderText } = render(
      <Formik initialValues={{ name: '' }} onSubmit={jest.fn()}>
        {() => (
          <div>
            <Settings />
          </div>
        )}
      </Formik>
    );
    expect(queryByText('Required')).toBeNull();
    fireEvent.focus(getByPlaceholderText('Detector interval'));
    fireEvent.blur(getByPlaceholderText('Detector interval'));
    expect(findByText('Required')).not.toBeNull();
  });
  test('shows error for invalid interval when toggling focus/blur', async () => {
    const { queryByText, findByText, getByPlaceholderText } = render(
      <Formik initialValues={{ name: '' }} onSubmit={jest.fn()}>
        {() => (
          <div>
            <Settings />
          </div>
        )}
      </Formik>
    );
    expect(queryByText('Required')).toBeNull();
    userEvent.type(getByPlaceholderText('Detector interval'), '-1');
    fireEvent.blur(getByPlaceholderText('Detector interval'));
    expect(findByText('Must be a positive integer')).not.toBeNull();
  });
  test('shows error for interval of 0 when toggling focus/blur', async () => {
    const { queryByText, findByText, getByPlaceholderText } = render(
      <Formik initialValues={{ name: '' }} onSubmit={jest.fn()}>
        {() => (
          <div>
            <Settings />
          </div>
        )}
      </Formik>
    );
    expect(queryByText('Required')).toBeNull();
    userEvent.type(getByPlaceholderText('Detector interval'), '0');
    fireEvent.blur(getByPlaceholderText('Detector interval'));
    expect(findByText('Must be a positive integer')).not.toBeNull();
  });
  test('shows error for empty window delay when toggling focus/blur', async () => {
    const { queryByText, findByText, getByPlaceholderText } = render(
      <Formik initialValues={{ name: '' }} onSubmit={jest.fn()}>
        {() => (
          <div>
            <Settings />
          </div>
        )}
      </Formik>
    );
    expect(queryByText('Required')).toBeNull();
    fireEvent.focus(getByPlaceholderText('Window delay'));
    fireEvent.blur(getByPlaceholderText('Window delay'));
    expect(findByText('Required')).not.toBeNull();
  });
  test('shows error for invalid window delay when toggling focus/blur', async () => {
    const { queryByText, findByText, getByPlaceholderText } = render(
      <Formik initialValues={{ name: '' }} onSubmit={jest.fn()}>
        {() => (
          <div>
            <Settings />
          </div>
        )}
      </Formik>
    );
    expect(queryByText('Required')).toBeNull();
    userEvent.type(getByPlaceholderText('Window delay'), '-1');
    fireEvent.blur(getByPlaceholderText('Window delay'));
    expect(findByText('Must be a non-negative integer')).not.toBeNull();
  });
});
