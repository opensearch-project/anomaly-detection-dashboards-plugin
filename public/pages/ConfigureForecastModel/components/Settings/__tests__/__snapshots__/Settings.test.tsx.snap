// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings /> spec renders the component 1`] = `
<div>
  <div
    class="euiFormRow euiFormRow--fullWidth euiFormRow--compressed"
    hint="How often the forecast runs to generate next value."
    id="random_html_id-row"
    title="Forecasting interval"
  >
    <div
      class="euiFormRow__labelWrapper"
    >
      <label
        aria-invalid="false"
        class="euiFormLabel euiFormRow__label"
        for="random_html_id"
      >
        <div
          style="line-height: 8px;"
        >
          <p>
            Forecasting interval
          </p>
          <br />
          <div
            class="euiText euiText--medium sublabel"
            style="max-width: 400px;"
          >
            How often the forecast runs to generate next value.
             
            <a
              class="euiLink euiLink--primary"
              href="https://opensearch.org/docs/latest/observing-your-data/forecast/index/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Learn more
              <svg
                aria-hidden="true"
                aria-label="External link"
                class="euiIcon euiIcon--small euiIcon-isLoading euiLink__externalIcon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                />
              </svg>
              <span
                class="euiScreenReaderOnly"
              >
                (opens in a new tab or window)
              </span>
            </a>
          </div>
        </div>
      </label>
    </div>
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        aria-describedby="random_html_id-help-0"
        class="euiFlexGroup euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
        id="random_html_id"
        style="max-width: 400px;"
      >
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <div
            style="width: 300px;"
          >
            <div
              class="euiFormControlLayout euiFormControlLayout--compressed"
            >
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <input
                  class="euiFieldNumber euiFieldNumber--compressed"
                  id="interval"
                  name="interval"
                  style="width: 100%;"
                  type="number"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <span
            class="euiBadge euiBadge--iconLeft unit-badge"
            style="background-color: rgb(211, 218, 230); color: rgb(0, 0, 0);"
            title="minutes"
          >
            <span
              class="euiBadge__content"
            >
              <span
                class="euiBadge__text"
              >
                minutes
              </span>
            </span>
          </span>
        </div>
      </div>
      <div
        class="euiFormHelpText euiFormRow__text"
        id="random_html_id-help-0"
      >
        The interval must be at least one minute.
      </div>
    </div>
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiFormRow euiFormRow--fullWidth euiFormRow--compressed"
    hint="Specify a window of delay for a forecaster to fetch data, if you need to account for extra processing time."
    id="random_html_id-row"
    title="Window delay"
  >
    <div
      class="euiFormRow__labelWrapper"
    >
      <label
        aria-invalid="false"
        class="euiFormLabel euiFormRow__label"
        for="random_html_id"
      >
        <div
          style="line-height: 8px;"
        >
          <p>
            Window delay
          </p>
          <br />
          <div
            class="euiText euiText--medium sublabel"
            style="max-width: 400px;"
          >
            Specify a window of delay for a forecaster to fetch data, if you need to account for extra processing time.
             
            <a
              class="euiLink euiLink--primary"
              href="https://opensearch.org/docs/latest/observing-your-data/forecast/index/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Learn more
              <svg
                aria-hidden="true"
                aria-label="External link"
                class="euiIcon euiIcon--small euiIcon-isLoading euiLink__externalIcon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                />
              </svg>
              <span
                class="euiScreenReaderOnly"
              >
                (opens in a new tab or window)
              </span>
            </a>
          </div>
        </div>
      </label>
    </div>
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        class="euiFlexGroup euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
        id="random_html_id"
        style="max-width: 400px;"
      >
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <div
            style="width: 300px;"
          >
            <div
              class="euiFormControlLayout euiFormControlLayout--compressed"
            >
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <input
                  class="euiFieldNumber euiFieldNumber--compressed"
                  id="windowDelay"
                  name="windowDelay"
                  style="width: 100%;"
                  type="number"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <span
            class="euiBadge euiBadge--iconLeft unit-badge"
            style="background-color: rgb(211, 218, 230); color: rgb(0, 0, 0);"
            title="minutes"
          >
            <span
              class="euiBadge__content"
            >
              <span
                class="euiBadge__text"
              >
                minutes
              </span>
            </span>
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiFormRow euiFormRow--fullWidth euiFormRow--compressed"
    hint="How far the forecast extends into the future."
    id="random_html_id-row"
    title="Horizon"
  >
    <div
      class="euiFormRow__labelWrapper"
    >
      <label
        aria-invalid="false"
        class="euiFormLabel euiFormRow__label"
        for="random_html_id"
      >
        <div
          style="line-height: 8px;"
        >
          <p>
            Horizon
          </p>
          <br />
          <div
            class="euiText euiText--medium sublabel"
            style="max-width: 400px;"
          >
            How far the forecast extends into the future.
             
            <a
              class="euiLink euiLink--primary"
              href="https://opensearch.org/docs/latest/observing-your-data/forecast/index/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Learn more
              <svg
                aria-hidden="true"
                aria-label="External link"
                class="euiIcon euiIcon--small euiIcon-isLoading euiLink__externalIcon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                />
              </svg>
              <span
                class="euiScreenReaderOnly"
              >
                (opens in a new tab or window)
              </span>
            </a>
          </div>
        </div>
      </label>
    </div>
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        aria-describedby="random_html_id-help-0"
        id="random_html_id"
        style="max-width: 400px;"
      >
        <div
          class="euiFlexGroup euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <div
              style="width: 300px;"
            >
              <div
                class="euiFormControlLayout euiFormControlLayout--compressed"
              >
                <div
                  class="euiFormControlLayout__childrenWrapper"
                >
                  <input
                    class="euiFieldNumber euiFieldNumber--compressed"
                    id="horizon"
                    max="180"
                    min="1"
                    style="width: 100%;"
                    type="number"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <span
              class="euiBadge euiBadge--iconLeft unit-badge"
              style="background-color: rgb(211, 218, 230); color: rgb(0, 0, 0);"
              title="intervals"
            >
              <span
                class="euiBadge__content"
              >
                <span
                  class="euiBadge__text"
                >
                  intervals
                </span>
              </span>
            </span>
          </div>
        </div>
      </div>
      <div
        class="euiFormHelpText euiFormRow__text"
        id="random_html_id-help-0"
      >
        Click the number field for more options.
      </div>
    </div>
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiFormRow euiFormRow--fullWidth euiFormRow--compressed"
    hint="How far back the model looks for training data."
    id="random_html_id-row"
    title="History"
  >
    <div
      class="euiFormRow__labelWrapper"
    >
      <label
        aria-invalid="false"
        class="euiFormLabel euiFormRow__label"
        for="random_html_id"
      >
        <div
          style="line-height: 8px;"
        >
          <p>
            History
          </p>
          <br />
          <div
            class="euiText euiText--medium sublabel"
            style="max-width: 400px;"
          >
            How far back the model looks for training data.
             
            <a
              class="euiLink euiLink--primary"
              href="https://opensearch.org/docs/latest/observing-your-data/forecast/index/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Learn more
              <svg
                aria-hidden="true"
                aria-label="External link"
                class="euiIcon euiIcon--small euiIcon-isLoading euiLink__externalIcon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                />
              </svg>
              <span
                class="euiScreenReaderOnly"
              >
                (opens in a new tab or window)
              </span>
            </a>
          </div>
        </div>
      </label>
    </div>
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        aria-describedby="random_html_id-help-0"
        class="euiFlexGroup euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
        id="random_html_id"
        style="max-width: 400px;"
      >
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <div
            style="width: 300px;"
          >
            <div
              class="euiFormControlLayout euiFormControlLayout--compressed"
            >
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <input
                  class="euiFieldNumber euiFieldNumber--compressed"
                  id="history"
                  max="10000"
                  min="40"
                  name="history"
                  style="width: 100%;"
                  type="number"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <span
            class="euiBadge euiBadge--iconLeft unit-badge"
            style="background-color: rgb(211, 218, 230); color: rgb(0, 0, 0);"
            title="intervals"
          >
            <span
              class="euiBadge__content"
            >
              <span
                class="euiBadge__text"
              >
                intervals
              </span>
            </span>
          </span>
        </div>
      </div>
      <div
        class="euiFormHelpText euiFormRow__text"
        id="random_html_id-help-0"
      >
        Minimum history: 40 intervals.
      </div>
    </div>
  </div>
</div>
`;
